src/PerfTreeParser/PerfTreeParser.class.st

```diff
@@ -165,25 +166,39 @@ PerfTreeParser >> numberOfSpacesAt: aNodeIndex [
 ]
 
 { #category : 'accessing' }
-PerfTreeParser >> percentageOf: aLine [
+PerfTreeParser >> percentageOf: aLineIndex [
 	"Gives the percentage from a function at a given line"
 
 	"In the case of multiples functions for one percentage, it gives the same percentage to every functions."
 
 	| stringLine numbers |
-	(aLine between: 1 and: self numberOfFunctions) ifFalse: [ ^ 0 ].
+	stringLine := self rawPercentageOf: aLineIndex.
 
-	stringLine := ((((self readLine: aLine) copyUpTo: $%) last: 5)
-		              copyWithout: $-) copyWithout: Character space.
-
-	stringLine = ((self readLine: aLine) last: stringLine size) ifTrue: [
-		^ self percentageOf: aLine - 1 ].
+	stringLine ifEmpty: [ ^ 0 ].
 
 	numbers := stringLine asNumber.
 
 	^ numbers
 ]
 
+{ #category : 'accessing' }
+PerfTreeParser >> rawPercentageOf: aLine [
+	"Gives the percentage from a function at a given line"
+
+	"In the case of multiples functions for one percentage, it gives the same percentage to every functions."
+
+	| stringLine |
+	(aLine between: 1 and: self numberOfFunctions) ifFalse: [ ^ '' ].
+
+	stringLine := ((((self readLine: aLine) copyUpTo: $%) last: 5)
+		               copyWithout: $-) copyWithout: Character space.
+
+	stringLine = ((self readLine: aLine) last: stringLine size) ifTrue: [
+		^ self rawPercentageOf: aLine - 1 ].
```

Comment:
Why do we need to re-parse?
The parser could remember the last percentage it saw, right?

Reply:
Unfortunately, as the parsing is done line by line. I don't think it is possible not to parse the line above again. 

However, it happens only if the line has no percentage, then it gets the one above until a line has a percentage written to it.

Reply:
I mean, the parser can remember the percentage of the previous line!
Actually, it's already parsed and stored in the node :)

Reply:
I'll try to do something about this